driver-core/src/main/com/mongodb/internal/authentication/AwsInternalCredential.java

```diff
@@ -0,0 +1,211 @@
+
+        if (awsCredentialSupplier != null) {
+            return obtainFromSupplier(startingCredential, awsCredentialSupplier);
+        }
+
+        BsonDocument ec2OrEcsResponse = null;
+
+        String accessKeyId;
+        if (startingCredential.getAccessKeyId() != null) {
+            accessKeyId = startingCredential.getAccessKeyId();
+        } else {
+            accessKeyId = System.getenv("AWS_ACCESS_KEY_ID");
+            if (accessKeyId == null) {
+                ec2OrEcsResponse = getEc2OrEcsResponse();
+                accessKeyId = ec2OrEcsResponse.getString("AccessKeyId").getValue();
+            }
+        }
+
+        String secretAccessKey;
+        if (startingCredential.getSecretAccessKey() != null) {
+            secretAccessKey = startingCredential.getSecretAccessKey();
+        } else {
+            if (System.getenv("AWS_SECRET_ACCESS_KEY") != null) {
+                secretAccessKey = System.getenv("AWS_SECRET_ACCESS_KEY");
+            } else {
+                ec2OrEcsResponse = ec2OrEcsResponse == null ? getEc2OrEcsResponse() : ec2OrEcsResponse;
+                secretAccessKey = ec2OrEcsResponse.getString("SecretAccessKey").getValue();
+            }
+        }
+
+        String sessionToken;
+        if (startingCredential.getAccessKeyId() != null) {
+            sessionToken = startingCredential.getSessionToken();
+        } else {
+            if (startingCredential.getSessionToken() != null) {
+                throw new IllegalArgumentException("A session token was provided without an access key identifier");
+            }
+            if ((System.getenv("AWS_SECRET_ACCESS_KEY") != null) || (System.getenv("AWS_ACCESS_KEY_ID") != null)
+                    || (System.getenv("AWS_SESSION_TOKEN") != null)) {
+                if (System.getenv("AWS_SECRET_ACCESS_KEY") == null || System.getenv("AWS_ACCESS_KEY_ID") == null) {
+                    throw new IllegalArgumentException("The environment variables 'AWS_ACCESS_KEY_ID' and 'AWS_SECRET_ACCESS_KEY' must "
+                            + "either both be set or both be null");
+                }
+                sessionToken = System.getenv("AWS_SESSION_TOKEN");
+            } else {
+                ec2OrEcsResponse = ec2OrEcsResponse == null ? getEc2OrEcsResponse() : ec2OrEcsResponse;
+                sessionToken = ec2OrEcsResponse.getString("Token").getValue();
+            }
+        }
+
+        return new AwsInternalCredential(accessKeyId, secretAccessKey, sessionToken);
+    }
+
+    private static AwsInternalCredential obtainFromSupplier(final AwsInternalCredential startingCredential, final Supplier<AwsCredential> awsCredentialSupplier) {
+        isTrueArgument("accessKeyId is null", startingCredential.getAccessKeyId() == null);
+        isTrueArgument("secret access key is null", startingCredential.getSecretAccessKey() == null);
+        isTrueArgument("session token is null", startingCredential.getSessionToken() == null);
```

Comment:
Session token can be null. User could theoretically provide a "regular" aws credential instead of a temporary one.

Reply:
This is just saying that if you provide a credential supplier, _nothing_ should have been provided in the connection string, e.g. `mongodb://localhost/?authMechanism=MONGODB-AWS`

Reply:
Or rather:

```
MongoCredential.createAwsCredential(null, null)
   .withMechanismProperty(AWS_SESSION_TOKEN_KEY, nulll)
   .withMechanismProperty(AWS_CREDENTIAL_PROVIDER_KEY, supplier)
```

Reply:
Thanks for the clarification.